In [2]:
%load_ext autoreload
%autoreload 2

In [3]:
import random
import sys
from itertools import chain

import numpy as np

from classifiers.basic_classifier import train_helper, BasicConvNet

In [4]:
data_dir = "data/bounding-box-obama-detector"
# data_dir = "data/obama-detector"

static_params = {
    "seed": 1,
    "data_dir": data_dir,
    "use_cuda": True,
    "val_batch_size": 64,
    "epochs": 10,
    "log_interval": 10
}

variable_params = {
    "lr": [0.1, 0.01, 0.001, 0.001],
    "momentum": [0.5, 0.6, 0.7, 0.8, 0.9],
    "batch_size": [32, 64, 128],
    "use_pretrained": [True, False],
    "feature_extract": [True, False]
}

In [4]:
# Run random search
used_variable_params = set()
param_space_size = np.prod([len(p) for p in variable_params])
while len(used_variable_params) < param_space_size:
    curr_params = {k: random.choice(v) for k, v in variable_params.items()}
    curr_variable_params = tuple(chain.from_iterable(curr_params.items()))

    if curr_variable_params not in used_variable_params:
        print("Current variable parameters:", curr_variable_params)
        used_variable_params.add(curr_variable_params)
        curr_params.update(static_params)
        train_helper(**curr_params)

Current variable parameters: ('lr', 0.001, 'momentum', 0.7, 'batch_size', 32, 'use_pretrained', True, 'feature_extract', True)
Params to learn:
	 classifier.6.weight
	 classifier.6.bias
Train Epoch: 1 [0/248 (0%)]	Loss: 0.692773

Test set: Average loss: 0.6690, Accuracy: 40/64 (62%)

Train Epoch: 2 [0/248 (0%)]	Loss: 0.694347

Test set: Average loss: 0.6630, Accuracy: 39/64 (61%)

Train Epoch: 3 [0/248 (0%)]	Loss: 0.732508

Test set: Average loss: 0.6538, Accuracy: 44/64 (69%)

Train Epoch: 4 [0/248 (0%)]	Loss: 0.709194

Test set: Average loss: 0.6519, Accuracy: 43/64 (67%)

Train Epoch: 5 [0/248 (0%)]	Loss: 0.707319

Test set: Average loss: 0.6558, Accuracy: 42/64 (66%)

Train Epoch: 6 [0/248 (0%)]	Loss: 0.711213

Test set: Average loss: 0.6543, Accuracy: 42/64 (66%)

Train Epoch: 7 [0/248 (0%)]	Loss: 0.610769

Test set: Average loss: 0.6546, Accuracy: 42/64 (66%)

Train Epoch: 8 [0/248 (0%)]	Loss: 0.632134

Test set: Average loss: 0.6506, Accuracy: 43/64 (67%)

Train Epoch: 9 [0/248 

Train Epoch: 6 [0/248 (0%)]	Loss: 0.664400

Test set: Average loss: 0.6873, Accuracy: 34/64 (53%)

Train Epoch: 7 [0/248 (0%)]	Loss: 0.714233

Test set: Average loss: 0.6821, Accuracy: 37/64 (58%)

Train Epoch: 8 [0/248 (0%)]	Loss: 0.711233

Test set: Average loss: 0.6922, Accuracy: 32/64 (50%)

Train Epoch: 9 [0/248 (0%)]	Loss: 0.702918

Test set: Average loss: 0.7064, Accuracy: 31/64 (48%)

Train Epoch: 10 [0/248 (0%)]	Loss: 0.736208

Test set: Average loss: 0.7369, Accuracy: 32/64 (50%)

Current variable parameters: ('lr', 0.1, 'momentum', 0.6, 'batch_size', 32, 'use_pretrained', True, 'feature_extract', True)
Params to learn:
	 classifier.6.weight
	 classifier.6.bias
Train Epoch: 1 [0/248 (0%)]	Loss: 0.692773

Test set: Average loss: 1.4422, Accuracy: 36/64 (56%)

Train Epoch: 2 [0/248 (0%)]	Loss: 1.640016

Test set: Average loss: 4.8612, Accuracy: 32/64 (50%)

Train Epoch: 3 [0/248 (0%)]	Loss: 3.609495

Test set: Average loss: 3.9841, Accuracy: 34/64 (53%)

Train Epoch: 4 [0/248 (

Train Epoch: 1 [0/248 (0%)]	Loss: 0.709296

Test set: Average loss: 4.6704, Accuracy: 32/64 (50%)

Train Epoch: 2 [0/248 (0%)]	Loss: 0.672187

Test set: Average loss: 1.5427, Accuracy: 33/64 (52%)

Train Epoch: 3 [0/248 (0%)]	Loss: 0.638275

Test set: Average loss: 1.3225, Accuracy: 33/64 (52%)

Train Epoch: 4 [0/248 (0%)]	Loss: 0.702932

Test set: Average loss: 0.7239, Accuracy: 36/64 (56%)

Train Epoch: 5 [0/248 (0%)]	Loss: 0.519329

Test set: Average loss: 0.6888, Accuracy: 43/64 (67%)

Train Epoch: 6 [0/248 (0%)]	Loss: 0.490576

Test set: Average loss: 0.7114, Accuracy: 39/64 (61%)

Train Epoch: 7 [0/248 (0%)]	Loss: 0.513170

Test set: Average loss: 0.7439, Accuracy: 39/64 (61%)

Train Epoch: 8 [0/248 (0%)]	Loss: 0.503807

Test set: Average loss: 0.8083, Accuracy: 35/64 (55%)

Train Epoch: 9 [0/248 (0%)]	Loss: 0.381869

Test set: Average loss: 0.8382, Accuracy: 40/64 (62%)

Train Epoch: 10 [0/248 (0%)]	Loss: 0.474625

Test set: Average loss: 0.8081, Accuracy: 40/64 (62%)

Current v

RuntimeError: CUDA out of memory. Tried to allocate 196.00 MiB (GPU 0; 10.91 GiB total capacity; 9.26 GiB already allocated; 34.19 MiB free; 591.51 MiB cached)

# Best results from Random Search
## Bounding Box Data
Basic CNN, both with 62% test accuracy:
```
('lr', 0.001, 'momentum', 0.9, 'batch_size', 32)  # epoch 2
('lr', 0.01, 'momentum', 0.6, 'batch_size', 32)  # epoch 5
```
With VGG:
```
('lr', 0.001, 'momentum', 0.7, 'batch_size', 32, 'use_pretrained', True, 'feature_extract', True)  # epoch 3
```


## Unfiltered Data
Both with 64% test accuracy:
```
('lr', 0.001, 'momentum', 0.8, 'batch_size', 32)  # epoch 10
('lr', 0.01, 'momentum', 0.7, 'batch_size', 64)  # epoch 7
```